lib/execution_engine2/db/MongoUtil.py

```diff
@@ -271,6 +271,42 @@ def check_if_already_finished(job_status):
             return True
         return False
 
+    def update_job_to_queued(
```

Comment:
This would need to return the status of the job in order to send the correct kafka message.
You could use the job object  if you wanted
J.queued = queue_time_now()
j.scheduler_id=scheduler_id
j.scheduler_type=condor

and then only set the status to queued if its "Created" or "Estimating" with the update_one statement

Reply:
As the code stands it doesn't set the queued timestamp if the job was set to cancelled, even though it was queued and now has a job id.

Reply:
The batch method doesn't - it just does https://github.com/kbase/execution_engine2/blob/a7a2d8a5359aefed16cc8ded49573c76b4837515/lib/execution_engine2/sdk/EE2Runjob.py#L325-L335

My implementation in the next PR does the same thing

Reply:
Right. It's kind of strange it it gets a scheduler id but doesn't have a queued timestamp, because how can you get a scheduler id without being queued?

Reply:
But having a scheduler job queued is not the same as when the job record entered the "queued" state. So depending on how you interpret what that timestamp means, that behavior could be correct or incorrect.

Reply:
> Right. It's kind of strange it it gets a scheduler id but doesn't have a queued timestamp, because how can you get a scheduler id without being queued?

That'll only happen if the job is set to running or canceled before Condor returns though. But if it is canceled, then you def want the scheduler info

Reply:
*if* the job is set to running prior to queued, we could add that other fix to `start_job` to set the queued time = running time, but from what you were saying it sounds like that should never happen under any reasonable circumstances

Reply:
>That'll only happen if the job is set to running or canceled before Condor returns though. But if it is canceled, then you def want the scheduler info so you can cancel it scheduler-side

The condor submission doesn't change the status of the job or not. You can submit a job to condor that is labeled any state. It will start running and immediately die though, but will take up a slot in the queue unless it was already submitted in the queue, as condor_rm cannot remove a job if you don't know the scheduler id yet. 

Maybe we should change the state of the job before submitting it and never submit a job unless its in the queued state. If it's in the canceled state, don't submit a job.

Reply:
Mmm, then you could wind up with jobs in the queued state that aren't queued.

I was thinking maybe we should add a `submit` time but that's basically the same as `created` other than maybe a few ms

Reply:
Well, not for batch jobs. There could be a significant difference there. But I don't think the added complexity would actually help.

Reply:
Should probably make a new ticket for this. Since it could take some time to submit a job to condor, this gives the opportunity for the fast cancel to at least prevent SOME of the jobs from actually being queued up in condor.

You wouldn't end up with jobs in the queued state that aren't queued because you would have to check the state multiple times.

Check state? Is it created? Then update to queued and add queued timestamp. Send kafka message.
Check state, is it queued? Then submit to condor, save scheduler id.

Reply:
> You can submit a job to condor that is labeled any state. It will start running and immediately die though

Let's say this happens:
* user submits job
* Condor submit starts
* User cancels job
* EE2 tries to update the job to the queued state, nothing happens

Now the job is queued on condor, and may stay queued for days or whatever, depending on how many jobs the user has scheduled. If the scheduler info is added to the job record, then the reaper thread could look for canceled jobs with jobs still in Condor and cancel them so they don't take a node (even for a short time).

Reply:
> Check state, is it queued? Then submit to condor...

If the cluster goes down here then you have a queued job record in Mongo but no job in Condor

Reply:
I kinda think we're overengineering here

Reply:
I'm pretty sure that's even more reason to mark the job as queued before submitting it.

Reply:
If mongo goes down, then the 14 day queued job reaper would clean it up.. Eventually.

Reply:
I don't understand that logic. If it's marked as created vs. queued, then that's accurate. There could even be code to recover the job on startup

Reply:
but again, I think we're solving problems that probably don't need to be solved

Reply:
Was referring to https://github.com/kbase/execution_engine2/pull/429#discussion_r772081171

Reply:
Oh, for that case the job was never really in the queued state as far as the user is concerned, so I think jumping from created->canceled is fine. 

We might want to have a internal state for the Condor job (like a boolean saying whether it was canceled or not or something) that the reaper could use to clean up Condor jobs

Reply:
The issue is that the thread is still going and submitting a bunch of cancelled jobs, because the condor job submission doesn't check the state of the job before submitting.

Reply:
The boolean is a good idea, if no job scheduler id is saved or  condor is temporarily unavailable we could save that.

Reply:
> The issue is that the thread is still going and submitting a bunch of cancelled jobs, because the condor job submission doesn't check the state of the job before submitting. If we check the state before submitting, and its created or queued, then we are good to go.

Oh I see, just a short circuit. We could do that right now by just pulling each job before the submit and checking it's in the created state to avoid a cancel race condition.

Of course it's still possible for the cancel to happen between that check and Condor returning, so it doesn't completely fix the problem, but it reduces it, esp. for large batches.

On the other hand, we could just leave it to the reaper thread, depending on how often it runs - add a `canceled_on_scheduler` (or something like that) boolean to the job, and for any jobs that are canceled but that isn't true have the reaper kill the Condor job

Reply:
yes exactly